pkg/controller/certificaterequests/ca/ca_test.go

```diff
@@ -391,3 +399,168 @@ func runTest(t *testing.T, test testT) {
 
 	test.builder.CheckAndFinish(err)
 }
+
+func TestCA_Sign(t *testing.T) {
+
+	rsaPair, err := pki.GenerateRSAPrivateKey(2048)
+	require.NoError(t, err)
+	rsaCSR := generateCSR(t, rsaPair)
+
+	tests := map[string]struct {
+		givenNamespace   string
+		givenCASecret    *corev1.Secret
+		givenCAIssuer    cmapi.GenericIssuer
+		givenCR          *cmapi.CertificateRequest
+		assertSignedCert func(t *testing.T, got *x509.Certificate)
+		wantErr          string
+	}{
+		"when the CertificateRequest has the duration field set, it should appear as notAfter on the signed ca": {
+			givenCAIssuer: gen.Issuer("issuer-1", gen.SetIssuerCA(cmapi.CAIssuer{
+				SecretName: "secret-1",
+			})),
+			givenCR: gen.CertificateRequest("cr-1",
+				gen.SetCertificateRequestCSR(rsaCSR),
+				gen.SetCertificateRequestIssuer(cmmeta.ObjectReference{
+					Name:  "issuer-1",
+					Group: certmanager.GroupName,
+					Kind:  "Issuer",
+				}),
+				gen.SetCertificateRequestDuration(&metav1.Duration{
+					Duration: 30 * time.Minute,
+				}),
+			),
+			givenCASecret: gen.SecretFrom(gen.Secret("secret-1"), gen.SetSecretNamespace("default"), gen.SetSecretData(secretDataFor(t, rsaPair,
+				&x509.Certificate{
+					SerialNumber: big.NewInt(1234),
+					IsCA:         true,
+				},
+			))),
+			givenNamespace: "default",
+			assertSignedCert: func(t *testing.T, got *x509.Certificate) {
+				assert.Equal(t, time.Now().UTC().Add(30*time.Minute), got.NotAfter)
```

Comment:
I get a mismatch here, am I misunderstanding the role of `cr.Spec.Duration`?

```sh
% go test ./pkg/controller/certificaterequests/ca/ -run TestCA_Sign
I0223 20:32:06.072826   16843 ca.go:128] cert-manager "msg"="certificate issued"
I0223 20:32:06.075193   16843 ca.go:128] cert-manager "msg"="certificate issued"
I0223 20:32:06.077904   16843 ca.go:128] cert-manager "msg"="certificate issued"
I0223 20:32:06.080078   16843 ca.go:128] cert-manager "msg"="certificate issued"
--- FAIL: TestCA_Sign (0.38s)
    --- FAIL: TestCA_Sign/when_the_CertificateRequest_has_the_duration_field_set,_it_should_appear_as_notAfter_on_the_signed_ca (0.00s)
        ca_test.go:440:
            	Error Trace:	ca_test.go:440
            	            				ca_test.go:545
            	Error:      	Not equal:
            	            	expected: time.Time{wall:0x459e058, ext:63749707326, loc:(*time.Location)(nil)}
            	            	actual  : time.Time{wall:0x0, ext:63749707326, loc:(*time.Location)(nil)}

            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,3 +1,3 @@
            	            	 (time.Time) {
            	            	- wall: (uint64) 72999000,
            	            	+ wall: (uint64) 0,
            	            	  ext: (int64) 63749707326,
            	Test:       	TestCA_Sign/when_the_CertificateRequest_has_the_duration_field_set,_it_should_appear_as_notAfter_on_the_signed_ca
FAIL
FAIL	github.com/jetstack/cert-manager/pkg/controller/certificaterequests/ca	0.411s
FAIL
```

Reply:
I think this is happening because time.Now() gives you a value to a higher precision than what ends up on the cert. I had the same issue when I was writing tests for the readiness controller and ended up rounding the `time.Now()` value to nearest 2 seconds.
I tried printing the cert that your tests generate and I see that NotAfter on the cert is `NotAfter:2021-02-24 13:37:53 +0000 UTC` whereas your expected NotAfter value is `2021-02-24 13:37:53.81280143 +0000 UTC`.

Reply:
Oohh thank you! I added `String()` in order to see a human-readable time, we can see the difference:

```
expected: 2021-02-24 14:26:38.587263 +0000 UTC
actual:   2021-02-24 14:26:38 +0000 UTC
```

I'll use your tip to fix the test case!

Reply:
Here is how I went:

```go
// The notAfter field uses a precision of 1 second; the current time is
// truncated before adding the certificate request's duration value.
assert.Equal(t, time.Now().UTC().Truncate(1*time.Second).Add(30*time.Minute).String(), got.NotAfter.String())
```

Reply:
Ah, yes truncating is obviously a better approach than rounding 😄 

`String()` makes it more readable, but maybe asserting the equality of the actual structs (rather than strings) is more important?

Reply:
OK, I'll do that then, and will show the .String() values in the error message instead

Reply:
That makes sense, I think